buildpacks/scala/src/errors.rs

```diff
@@ -0,0 +1,274 @@
+use indoc::formatdoc;
+use libcnb::Error;
+use libherokubuildpack::log::log_error;
+use std::ffi::OsString;
+use std::fmt::Debug;
+use std::process::ExitStatus;
+
+#[derive(Debug)]
+pub enum ScalaBuildpackError {
+    CouldNotWriteSbtExtrasScript(std::io::Error),
+    CouldNotSetExecutableBitForSbtExtrasScript(std::io::Error),
+    CouldNotWriteSbtWrapperScript(std::io::Error),
+    CouldNotSetExecutableBitForSbtWrapperScript(std::io::Error),
+    MissingSbtBuildPropertiesFile,
+    SbtPropertiesFileReadError(std::io::Error),
+    InvalidSbtPropertiesFile(java_properties::PropertiesError),
+    MissingDeclaredSbtVersion,
+    UnsupportedSbtVersion(String),
+    SbtVersionNotInSemverFormat(String, semver::Error),
+    SbtBuildIoError(std::io::Error),
+    SbtBuildUnexpectedExitCode(ExitStatus),
+    SbtInstallIoError(std::io::Error),
+    SbtInstallUnexpectedExitCode(ExitStatus),
+    CouldNotWriteSbtPlugin(std::io::Error),
+    NoBuildpackPluginAvailable(String),
+    CouldNotParseBooleanFromProperty(String, std::str::ParseBoolError),
+    CouldNotParseBooleanFromEnvironment(String, std::str::ParseBoolError),
+    CouldNotParseListConfigurationFromProperty(String, shell_words::ParseError),
+    CouldNotParseListConfigurationFromEnvironment(String, shell_words::ParseError),
+    CouldNotConvertEnvironmentValueIntoString(String, OsString),
+    CouldNotReadSbtOptsFile(std::io::Error),
+    CouldNotParseListConfigurationFromSbtOptsFile(shell_words::ParseError),
+    MissingStageTask,
+    AlreadyDefinedAsObject,
+}
+
+#[allow(clippy::too_many_lines)]
+pub fn log_user_errors(error: ScalaBuildpackError) {
+    match error {
+        ScalaBuildpackError::MissingDeclaredSbtVersion |
+        ScalaBuildpackError::MissingSbtBuildPropertiesFile => log_error(
+            "No sbt version defined",
+            formatdoc! { "
+                Your scala project must include project/build.properties and define a value for
+                the `sbt.version` property.
+            " },
+        ),
+
+        ScalaBuildpackError::UnsupportedSbtVersion(version) => log_error(
+            "Unsupported sbt version",
+            formatdoc! { "
+                You have defined an unsupported `sbt.version` ({version}) in the project/build.properties
+                file. You must use a version of sbt between 0.11.0 and 1.x.
+            ", version = version },
+        ),
+
+        ScalaBuildpackError::SbtBuildIoError(error) => log_error(
+            "Failed to build app with sbt",
+            formatdoc! { "
+                We're sorry this build is failing!  If you can't find the issue in application code,
+                please submit a ticket so we can help: https://help.heroku.com/
```

Comment:
It sounds counterintuitive but I highly recommend against linking to our generic help docs or encouraging people to open tickets. 

When they see this they're likely to think that it's an internal error that they cannot fix and open a ticket when the problem is might be coming from their application config. Same with telling people to open issues. People who have never used sbt before in their life will be following the output naively, open an issue, then get mad when you don't fix their app for them. I would also stay away from "Sorry" as it has the connotation of taking on responsibility. It's endearing, but the 100,000th time a developer sees it they won't be endeared. 

Some other general guidelines: Make statements that are provably true (like "Running sbt task failed").  If you're guessing about something, state why you're guessing. Don't imply that an action will likely fix their problem unless you're extremely confident about it (also state why you're confident). Prefer statements over questions i.e. "Help:" instead of "Did you mean?".

FWIW these generic standard IO errors are the reason why I wrote `which_program`.  I haven't actually incorporated it yet, so I don't have feedback if it's a good idea or not.

Maybe we could say something like:

```
Running sbt task failed

This failure occurred before the executable was invoked. Details are below. If this error persists, 
try deploying the Heroku Scala Getting Started Guide to a new application as a debugging step:

https://devcenter.heroku.com/articles/getting-started-with-scala

Details: {details}
```

I'm not sure if we should actually encourage people to do that as a debugging step (or perhaps give them more info/context on what we mean by that). It's an idea though.

Reply:
i'll follow up with you around the best practices for errors here @schneems